ext/src/helpers/symbol_enum.rs

```diff
@@ -0,0 +1,64 @@
+use super::static_id::StaticId;
+use magnus::{exception::arg_error, Error, Symbol, Value};
+use std::fmt::Display;
+
+/// Represents an enum as a set of Symbols (using `StaticId`).
+/// Each symbol maps to a value of the enum's type.
+pub struct SymbolEnum<'a, T: Clone> {
+    what: &'a str,
+    mapping: Mapping<T>,
+}
+
+impl<'a, T: Clone> SymbolEnum<'a, T> {
+    pub fn new(what: &'a str, mapping: Vec<(StaticId, T)>) -> Self {
+        Self {
+            what,
+            mapping: Mapping(mapping),
+        }
+    }
+
+    /// Map a Magnus `Value` to a Value in the Enum.
+    /// Returns an ArgumentError listing all possible valid symbols
+    /// when the passed in `Value` isn't one of the symbol.
+    pub fn get(&self, needle: Value) -> Result<T, magnus::Error> {
+        let needle: Symbol = needle.try_convert().map_err(|_| self.error(needle))?;
+        let id = magnus::value::Id::from(needle);
+
+        self.mapping
+            .0
+            .iter()
+            .find(|(haystack, _)| *haystack == id)
```

Comment:
In my experimentation this was faster than a HashMap, which isn't too surprising given our enums are small.